### ISSUE TITLE
Subnets with purpose set to PRIVATE_SERVICE_CONNECT are broken in v7.3.0

### ISSUE THREAD

**Issue by colereynolds** [terraform-google-modules/terraform-google-network#487]
### TL;DR

After upgrading to v7.3.0 from v7.2.0 and running a plan, existing subnets that have `purpose` set to `PRIVATE_SERVICE_CONNECT` need to be replaced. This is not an issue with v7.2.0, which we used to create these subnets. We have a workaround of pinning to v7.2.0 until this is fixed.

### Expected behavior

The plan would show no changes to existing subnets that have `purpose` set to `PRIVATE_SERVICE_CONNECT`, who's configuration has not changed.

### Observed behavior

```
# module.[redacted].module.main.module.subnets.google_compute_subnetwork.subnetwork["us-central1/[redacted]"] must be replaced
-/+ resource "google_compute_subnetwork" "subnetwork" {
      ~ creation_timestamp         = "2023-08-04T13:34:55.738-07:00" -> (known after apply)
      + external_ipv6_prefix       = (known after apply)
      + fingerprint                = (known after apply)
      ~ gateway_address            = "[redacted]" -> (known after apply)
      ~ id                         = "projects/[redacted]/regions/us-central1/subnetworks/[redacted]" -> (known after apply)
      + ipv6_cidr_range            = (known after apply)
        name                       = "[redacted]"
      ~ network                    = "https://www.googleapis.com/compute/v1/projects/[redacted]/global/networks/[redacted]" -> "[redacted]"
      ~ private_ipv6_google_access = "DISABLE_GOOGLE_ACCESS" -> (known after apply)
      ~ purpose                    = "PRIVATE" -> "PRIVATE_SERVICE_CONNECT" # forces replacement
      ~ self_link                  = "https://www.googleapis.com/compute/v1/projects/[redacted]/regions/us-central1/subnetworks/[redacted]" -> (known after apply)
      ~ stack_type                 = "IPV4_ONLY" -> (known after apply)
        # (6 unchanged attributes hidden)

      ~ log_config {
          - metadata_fields      = [] -> null
            # (4 unchanged attributes hidden)
        }
    }
```

### Terraform Configuration

```hcl
#########################################################
## module that contains our VPC configuration
#########################################################
module "[redacted]_vpc" {
  source                                 = "../../modules/base_vpc" # see below
  project_id                             = local.base_net_hub_project_id
  environment_code                       = "[redacted]"
  org_id                                 = var.org_id
  parent_folder                          = var.parent_folder
  bgp_asn_subnet                         = local.bgp_asn_number
  default_region1                        = var.default_region1
  default_region2                        = var.default_region2
  domain                                 = var.domain
  dns_enable_inbound_forwarding          = var.base_hub_dns_enable_inbound_forwarding
  dns_enable_logging                     = var.base_hub_dns_enable_logging
  firewall_enable_logging                = var.base_hub_firewall_enable_logging
  nat_enabled                            = var.base_hub_nat_enabled
  nat_bgp_asn                            = var.base_hub_nat_bgp_asn
  nat_num_addresses_region1              = var.base_hub_nat_num_addresses_region1
  nat_num_addresses_region2              = var.base_hub_nat_num_addresses_region2
  windows_activation_enabled             = var.base_hub_windows_activation_enabled
  folder_prefix                          = var.folder_prefix
  delete_default_internet_gateway_routes = "false"
  allow_all_egress_ranges                = local.allow_all_egress_ranges
  allow_all_ingress_ranges               = local.allow_all_ingress_ranges
  enable_shared_vpc_host                 = true

  private_service_cidr = [
    {
      name  = "[redacted]"
      range = "[redacted]"
    },
    {
      name  = "[redacted]"
      range = "[redacted]"
    },
    {
      name  = "[redacted]"
      range = "[redacted]"
    },
    {{
      name  = "[redacted]"
      range = "[redacted]"
    }
  ]

  subnets = [
    {
      subnet_name           = "sb-[redacted]-psc-${var.default_region1}"
      subnet_ip             = local.[redacted]_subnet_primary_ranges["psc_region1"]
      subnet_region         = var.default_region1
      subnet_private_access = "true"
      purpose               = "PRIVATE_SERVICE_CONNECT"
      subnet_flow_logs      = var.subnetworks_enable_logging
      description           = "[redacted]"
    },
    {
      subnet_name           = "sb-[redacted]-psc-${var.default_region2}"
      subnet_ip             = local.[redacted]_subnet_primary_ranges["psc_region2"]
      subnet_region         = var.default_region2
      subnet_private_access = "true"
      purpose               = "PRIVATE_SERVICE_CONNECT"
      subnet_flow_logs      = var.subnetworks_enable_logging
      description           = "[redacted]"
    }
  ]
  secondary_ranges = {}
}

#########################################################
## module that calls the terraform-google-network module
#########################################################
module "main" {
  source                                 = "terraform-google-modules/network/google"
  version                                = "~> 7.3.0"
  project_id                             = var.project_id
  network_name                           = local.network_name
  shared_vpc_host                        = var.enable_shared_vpc_host ? true : false
  delete_default_internet_gateway_routes = var.delete_default_internet_gateway_routes
  routing_mode                           = var.routing_mode

  subnets          = var.subnets
  secondary_ranges = var.secondary_ranges

  routes = concat(
    [{
      name              = "rt-${local.vpc_name}-1000-all-default-private-api"
      description       = "Route through IGW to allow private google api access."
      destination_range = "199.36.153.8/30"
      next_hop_internet = "true"
      priority          = "1000"
    }],
    var.nat_enabled ?
    [
      {
        name              = "rt-${local.vpc_name}-1000-egress-internet-default"
        description       = "Tag based route through IGW to access internet"
        destination_range = "0.0.0.0/0"
        tags              = "egress-internet"
        next_hop_internet = "true"
        priority          = "1000"
      }
    ]
    : [],
    var.windows_activation_enabled ?
    [{
      name              = "rt-${local.vpc_name}-1000-all-default-windows-kms"
      description       = "Route through IGW to allow Windows KMS activation for GCP."
      destination_range = "35.190.247.13/32"
      next_hop_internet = "true"
      priority          = "1000"
      }
    ]
    : []
  )
}
```

### Terraform Version

```sh
Terraform v1.4.5
on darwin_arm64
+ provider registry.terraform.io/hashicorp/google v4.78.0
+ provider registry.terraform.io/hashicorp/google-beta v4.78.0
+ provider registry.terraform.io/hashicorp/random v3.5.1
```

### Additional information

I reviewed the changes to the subnets submodule between v7.3.0 and v7.2.0 and the breaking change appears to have come from this [commit](https://github.com/terraform-google-modules/terraform-google-network/commit/71a11777cb5c432459ae0cdca2bc625c5c513651#diff-62243b45b9895d5deeee758deb73d041ec4c837c528532d7af92a72596d68f22) to modules/subnets/variables.tf. If I comment out [line 41](https://github.com/terraform-google-modules/terraform-google-network/blob/71a11777cb5c432459ae0cdca2bc625c5c513651/modules/subnets/variables.tf#L41), which removes the `purpose` argument from the object block, Terraform no longer wants to recreate the resource. I can't explain this behavior but I'm hoping somewhere that understands the backend APIs a little better will and can implement a fix.

**Comment by colereynolds** [comment]
Update: This may be a backend issue. I opened up a support case and they are investigating why the purpose changed on two subnets that I deployed a week ago.

**Comment by imrannayer** [comment]
@colereynolds if it is a provider issue can you plz open issue on [TPG github](https://github.com/hashicorp/terraform-provider-google/issues)?

**Comment by colereynolds** [comment]
@imrannayer I opened this thinking that it was a module logic issue but it was indeed a backend issue that the product team is investigating. Please disregard.
